crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -0,0 +1,439 @@
+use super::{
+    index_allocator::{ModuleAffinityIndexAllocator, SlotId},
+    MemoryAllocationIndex,
+};
+use crate::{
+    CompiledModuleId, InstanceAllocationRequest, Memory, MemoryImageSlot, Mmap,
+    PoolingInstanceAllocatorConfig,
+};
+use anyhow::{anyhow, bail, Context, Result};
+use libc::c_void;
+use std::sync::Mutex;
+use wasmtime_environ::{
+    DefinedMemoryIndex, MemoryPlan, MemoryStyle, Module, Tunables, WASM_PAGE_SIZE,
+};
+
+/// Represents a pool of WebAssembly linear memories.
+///
+/// A linear memory is divided into accessible pages and guard pages.
+///
+/// Each instance index into the pool returns an iterator over the base
+/// addresses of the instance's linear memories.
+///
+/// A diagram for this struct's fields is:
+///
+/// ```ignore
+///                       memory_size
+///                           /
+///         max_accessible   /                    memory_and_guard_size
+///                 |       /                               |
+///              <--+--->  /                    <-----------+---------->
+///              <--------+->
+///
+/// +-----------+--------+---+-----------+     +--------+---+-----------+
+/// | PROT_NONE |            | PROT_NONE | ... |            | PROT_NONE |
+/// +-----------+--------+---+-----------+     +--------+---+-----------+
+/// |           |<------------------+---------------------------------->
+/// \           |                    \
+/// mapping     |     `max_instances * max_memories` memories
+///            /
+///    initial_memory_offset
+/// ```
+#[derive(Debug)]
+pub struct MemoryPool {
+    mapping: Mmap,
+    index_allocator: ModuleAffinityIndexAllocator,
+    // If using a copy-on-write allocation scheme, the slot management. We
+    // dynamically transfer ownership of a slot to a Memory when in
+    // use.
+    image_slots: Vec<Mutex<Option<MemoryImageSlot>>>,
+    // The size, in bytes, of each linear memory's reservation, not including
+    // any guard region.
+    memory_size: usize,
+    // The size, in bytes, of each linear memory's reservation plus the trailing
+    // guard region allocated for it.
+    memory_and_guard_size: usize,
+    // The maximum size that can become accessible, in bytes, of each linear
+    // memory. Guaranteed to be a whole number of wasm pages.
+    max_accessible: usize,
+    // The size, in bytes, of the offset to the first linear memory in this
+    // pool. This is here to help account for the first region of guard pages,
+    // if desired, before the first linear memory.
+    initial_memory_offset: usize,
+    // The maximum number of memories that can be allocated concurrently, aka
+    // our pool's capacity.
+    max_total_memories: usize,
+    // The maximum number of memories that a single core module instance may
+    // use.
+    memories_per_instance: usize,
```

Comment:
add comment about how this is needed for validation but does not factor into the pre-allocated pool size and its layout